testing/test_config.py

```diff
@@ -1176,3 +1176,36 @@ def pytest_addoption(parser):
         ["*pytest*{}*imported from*".format(pytest.__version__)]
     )
     assert result.ret == EXIT_USAGEERROR
+
+
+def test_report_implicit_args(testdir, monkeypatch):
+    p = testdir.makepyfile(
+        """
+            def test():
+                pass
+        """
+    )
+    result = testdir.runpytest(str(p))
+    assert "implicit args:" not in result.stdout.str()
+
+    ini = testdir.makeini(
+        """
+        [pytest]
+        addopts = -k "foo and bar"
+    """
+    )
+    result = testdir.runpytest(str(p))
+    result.stdout.fnmatch_lines(
+        ["implicit args: \"-k 'foo and bar'\" (addopts config)"]
+    )
+
+    monkeypatch.setenv("PYTEST_ADDOPTS", "-v")
+    result = testdir.runpytest(str(p))
+    result.stdout.fnmatch_lines(
+        ["implicit args: \"-k 'foo and bar'\" (addopts config), '-v' (PYTEST_ADDOPTS)"]
+    )
+
+    ini.remove()
+    monkeypatch.setenv("PYTEST_ADDOPTS", "-v -k 'bar'")
+    result = testdir.runpytest(str(p))
+    result.stdout.fnmatch_lines(["implicit args: '-v -k bar' (PYTEST_ADDOPTS)"])
```

Comment:
Maybe unit-tests would be better/faster here, except for one end-to-end, but waiting for general feedback first.